src/System.Private.CoreLib/shared/System/Number.DiyFp.cs

```diff
@@ -8,121 +8,141 @@ namespace System
 {
     internal static partial class Number
     {
-        // An exteneded floating-point data structure which is required by Grisu3 algorithm.
-        // It defines a 64-bit significand and a 32-bit exponent,
-        // which is EXTENDED compare to IEEE double precision floating-point number (53 bits significand and 11 bits exponent).
-        //
-        // Original Grisu algorithm produces suboptimal results. To shorten the output (which is part of Grisu2/Grisu3's job),
-        // we need additional 11 bits of the significand f and larger exponent e (A larger exponent range is used to avoid overflow. A 32-bit exponent is far big enough).
-        // To fully understand how Grisu3 uses this data structure to get better result, please read http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf
-        internal struct DiyFp
+        // This is a port of the `DiyFp` implementation here: https://github.com/google/double-conversion/blob/a711666ddd063eb1e4b181a6cb981d39a1fc8bac/double-conversion/diy-fp.h
+        // The backing structure and how it is used is described in more detail here: http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf
+
+        // This "Do It Yourself Floating Point" class implements a floating-point number with a ulong significand and an int exponent.
+        // Normalized DiyFp numbers will have the most significant bit of the significand set.
+        // Multiplication and Subtraction do not normalize their results.
+        // DiyFp are not designed to contain special doubles (NaN and Infinity).
+        internal readonly ref struct DiyFp
```

Comment:
What's the advantage of being a `ref struct` here?

Reply:
Only the guarantee that it has to stay on the stack.

The fact that this is on the stack and is readonly is relied on for several of the perf optimizations that have been made elsewhere, so making it `readonly ref` asserts the behavior isn't changed.